### ISSUE TITLE
SSH and HTTPS Support?

### ISSUE THREAD

**Issue by FantaBlueMystery** [NginxProxyManager/nginx-proxy-manager#344]
Can you add SSH Support? (Nice to have)

For example by my SSH-Proxy 
```
stream {

    upstream ssh {
        server 10.8.0.16:22;
    }

    upstream https {
        server 10.8.0.15:443;
    }

    upstream https2 {
        server 10.8.0.18:443;
    }

    map $ssl_preread_server_name $name {
     mysdomain.com https2;
      default https;
    }

    map $ssl_preread_protocol $upstream {
        default ssh;
        "TLSv1.2" $name;
        "TLSv1.3" $name;
        "TLSv1.1" $name;
        "TLSv1.0" $name;
    }

    # SSH and SSL on the same port
    server {
        listen 443;

        proxy_pass $upstream;
        ssl_preread on;
    }
```

**Comment by Thijmen** [comment]
I am already forwarding SSH with streams, would this help you?

**Comment by FantaBlueMystery** [comment]
Hi there
Thanks for the answer! Your answer would only be a part
because according to the protocol "$ ssl_preread_protocol" it would have to be divided into "TLSv1.2", "TLSv1.3", "TLSv1.1", "TLSv1.0" and the "default" for the ssh.
All this would be necessary to separate the HTTPS and SSH on one port (443). If it could then be separated by domain, that would be wonderful.

Would that be possible with the streams in the UI? Can you set the stream again on port 443? It would be important that only one port is present on the outside.

**Comment by Thijmen** [comment]
Is that even possible, @FantaBlueMystery ?

**Comment by FantaBlueMystery** [comment]
yes clearly see my first post, currently I have another nginx before npm that uses this config and on 'server 10.8.0.15:443;' forward to the npm 

:)

**Comment by gustavosbarreto** [comment]
Hi @FantaBlueMystery, did you know ShellHub? I think you can configure NGINX Proxy Manager to work together ShellHub for SSH access.

I'm doing some tests with NGINX Proxy Manager and ShellHub to provide HTTPS access.

**Comment by FantaBlueMystery** [comment]
Hello @gustavosbarreto thanks for the info, I didn't know "shellhub" until now. It looks very interesting e.g. with "web-based user interface".

But here it would also be the case that it should be accessible using the SSH protocol. As an example, we would use sftp or the Ansible program.

But I'll still have a look at "shellhub". But I think to get the destination a port on the router you still can't get without the settings like in my first post.

Port 22 (or another ssh port) should not be visible from the outside. Therefore, a camouflage via https (443) is ingenious, since you can determine the HTTPs server and SSH in the internal network at the same time based on the domain (subdomaining).

**Comment by ursus69** [comment]
Has there been any advancement on this enhancement?
It would make it the ultimate NGinx project if we could make this happen!
almost all the businesses blocks other ports than 80 and 443, so this will allow me to connect to my home undetected.
If we could also add other streams like VPN over 443 it will be great!
Cordially

**Comment by ursus69** [comment]
> yes clearly see my first post, currently I have another nginx before npm that uses this config and on 'server 10.8.0.15:443;' forward to the npm
> 
> :)

How do you achieve this? which docker image you use for your head Nginx with streams?
Cheers!

**Comment by FantaBlueMystery** [comment]
Hey @ursus69 sorry for my late answer:

docker-compose.yml
```version: "3"
services:
  npm:
    image: nginx
    container_name: sshproxy
    restart: always
    ports:
      - 444:443
      - 84:80
    volumes:
      - ./nginx.conf:/etc/nginx/nginx.conf:ro
      - ./logs/:/var/log/nginx/
```

Edit your nginx.conf and add:

> Can you add SSH Support? (Nice to have)
> 
> For example by my SSH-Proxy
> 
> ```
> stream {
> 
>     upstream ssh {
>         server 10.8.0.16:22;
>     }
> 
>     upstream https {
>         server 10.8.0.15:443;
>     }
> 
>     upstream https2 {
>         server 10.8.0.18:443;
>     }
> 
>     map $ssl_preread_server_name $name {
>      mysdomain.com https2;
>       default https;
>     }
> 
>     map $ssl_preread_protocol $upstream {
>         default ssh;
>         "TLSv1.2" $name;
>         "TLSv1.3" $name;
>         "TLSv1.1" $name;
>         "TLSv1.0" $name;
>     }
> 
>     # SSH and SSL on the same port
>     server {
>         listen 443;
> 
>         proxy_pass $upstream;
>         ssl_preread on;
>     }
> ```

Simple and fast :)

**Comment by ursus69** [comment]
Many thanks @FantaBlueMystery, I t works like a charm!

**Comment by fdelucchijr** [comment]
This is an exiting enhancement!!! It would add an extra layer of security for my network! 
Praying to see it in NPM soon!

**Comment by litio2001** [comment]
Do you know when this new enhancement will be available in the docker NPM image?

thanks in advance for the wonderful development!!

**Comment by chaptergy** [comment]
I don't think this code snippet will ever be officially added to NPM unless someone creates a pull request adding this and creating a user interface where this is configurable. As it is now, this is very specific for one ssh server, and the file would have to be edited manually anyways, so there is no point in adding it. You would have to add this nginx config inside your container yourself.

**Comment by fdelucchijr** [comment]
> 
> 
> I don't think this code snippet will ever be officially added to NPM unless someone creates a pull request adding this and creating a user interface where this is configurable. As it is now, this is very specific for one ssh server, and the file would have to be edited manually anyways, so there is no point in adding it. You would have to add this nginx config inside your container yourself.

This comment make me kind of sad, but its acceptable. I get an alternative.
Using Cloudflare's Argo Tunnels make this so easy! also provides TLS secure connection, IP hidding, DDNS, CGNAT Bypass and other amazing features.

For ssh with WebRender you can use this guide: [This](https://github.com/aeleos/cloudflared#enabling-ssh-access-via-web-rendered-terminal)

To use the setup with the ssh cli you can use [this](https://developers.cloudflare.com/cloudflare-one/tutorials/ssh#connect-from-a-client-machine)

And, of you think this offers you a ProxyJump to far away when your locally you can use [this amazing blog](https://mike.place/2017/ssh-match/)

KEEP IN MIND that this heavely relies in Cloudflare/LetsEncrypt DNS certificates and (in this instruction) Docker.

**Comment by stefanwerfling** [comment]
Hello everyone, I'm a big fan of NPM!

I've started to write my own software for this implementation and I'm still looking for testers, whoever is interested can take a look at the whole thing.

I'm just posting this because the idea isn't being picked up. The whole thing is still under development. I use it for myself already in a live test. You can find it through my profile.

**Comment by marky421** [comment]
Can this feature be used to enable SSH support? Would the streams feature come into play? I'm hoping to only have ports 80/443 open on my firewall but I want to be able to SSH and host some https sites. 

https://nginxproxymanager.com/advanced-config/#custom-nginx-configurations

**Comment by stefanwerfling** [comment]
@marky421 I do not fully understand the question.

The example above shows how an SSH protocol can be placed over the HTTPS port.
Nginx looks at which protocol applies to TLS and directs it to the appropriate upstream.
If none of the protocols match, the default target is the SSH server.
The whole thing only works because TLS AND SSH descend from SSL.

Magic :)

**Comment by albeec13** [comment]
@

> Can this feature be used to enable SSH support? Would the streams feature come into play? I'm hoping to only have ports 80/443 open on my firewall but I want to be able to SSH and host some https sites.
> 
> https://nginxproxymanager.com/advanced-config/#custom-nginx-configurations

See my reply in the linked thread here: https://github.com/NginxProxyManager/nginx-proxy-manager/issues/646#issuecomment-2613387551

tl;dr is you can use `/data/nginx/custom/stream.conf` to accomplish what you want, but only if you have NO other proxy hosts using 443 defined in the standard NPM GUI. You can define them usinig the OP's `ssl_preread_server_name` method above, but I'm not sure if that and the `upstream` directives are sufficient for getting ssl certs and such working for an HTTPS server without a standard `http{ server { ... } }` nginx block/directive.

**Comment by stefanwerfling** [comment]
@albeec13 Why are you reposting an answer that has been clarified for a long time? Your example, which came later, explains the same context. 

It's clear that if npm doesn't build an interface for this, this issue will only be closed if it is rejected. That's okay too. Then you can put your own config or a pre-nginx in front of it.

:)

**Comment by albeec13** [comment]
> [@albeec13](https://github.com/albeec13) Why are you reposting an answer that has been clarified for a long time? Your example, which came later, explains the same context.
> 
> It's clear that if npm doesn't build an interface for this, this issue will only be closed if it is rejected. That's okay too. Then you can put your own config or a pre-nginx in front of it.
> 
> :)

I was running into a similar problem and decided to respond to this existing thread instead of starting a new one. Moreover, there were two important distinctions between my response and others I've seen which are:

1. You can accomplish ssh/tls at the same time on a _single server_ without needing to run another instance of nginx 
2. If you decide to use `./custom/stream.conf`, you will not be able to use other proxy hosts on 443 from the normal GUI route in NPM

I felt these were important/useful to the next person who may run into this issue. If it's not of use to you, no need to snark about it.

**Comment by stefanwerfling** [comment]
@albeec13 Okay, I understand, then we got off on the wrong foot. 

> I'm not sure if that and the upstream directives are sufficient for getting ssl certs and such working for an HTTPS server without a standard http{ server { ... } } nginx block/directive.

Apart from npm, this is possible and this is also what my app is aiming for. 

As with the 443 port, you also carry out “domain splitting” on the 80 port, i.e. a map to different upstreams. For example, all domains that point to the 80 port are directed to the upstream 127.0.0.1:10080. 

There the 10080 waits for a response depending on the domain with 
```
server {
    listen 10080;
    ...
}

```
each of these "servers" has a location /.well-know/ entry. 

Now you could ask yourself how the original client IP gets to the server. 

This is also easy, in the 
```
stream {
   server {
      ...
      proxy_protocol on;
   }
}
```

"proxy_protocol" is activated and on the 

```
http {
   server {
      listen 10080 proxy_protocol;
   }
}
```

Activated. Now the client IPs will also be passed on.

**Comment by albeec13** [comment]
> [@albeec13](https://github.com/albeec13) Okay, I understand, then we got off on the wrong foot.

No worries!

> 
> > I'm not sure if that and the upstream directives are sufficient for getting ssl certs and such working for an HTTPS server without a standard http{ server { ... } } nginx block/directive.
> 
> Apart from npm, this is possible and this is also what my app is aiming for.

I'll have to check that out, thanks!

> 
> As with the 443 port, you also carry out “domain splitting” on the 80 port, i.e. a map to different upstreams. For example, all domains that point to the 80 port are directed to the upstream 127.0.0.1:10080.
> 
> There the 10080 waits for a response depending on the domain with
> 
> ```
> server {
>     listen 10080;
>     ...
> }
> ```
> 
> each of these "servers" has a location /.well-know/ entry.
> 
> Now you could ask yourself how the original client IP gets to the server.
> 
> This is also easy, in the
> 
> ```
> stream {
>    server {
>       ...
>       proxy_protocol on;
>    }
> }
> ```
> 
> "proxy_protocol" is activated and on the
> 
> ```
> http {
>    server {
>       listen 10080 proxy_protocol;
>    }
> }
> ```
> 
> Activated. Now the client IPs will also be passed on.

Cool, I was not aware of the proxy_protocol directive, I'll need to look into that some more. Thanks!

**Comment by stefanwerfling** [comment]
@komodikkio 

I've created an overview of my route for you. If no suitable domain is specified, your browser will default to "ssh," which it won't be able to do (it won't understand because it's an SSH protocol). "https" upstream is empty and cannot be reached.

![Image](https://github.com/user-attachments/assets/10ca53f3-ce8d-4fba-a57e-67abb75a14b3)

On the upstream "ssh," I see that you're specifying port "8443," which not sounds like the SSH server. Please check this again; the normal SSH port is 22. If there's a web server behind 8443, it won't understand SSH and will respond with a "Bad Request."

I hope this helps you.

**Comment by komodikkio** [comment]
Hi Stefan, thanks for the provided oveview.
What is failing to me it's the ssh connection when i call, from public, the NPM's port 443
When i try to ssh to  mypublic ip:443 i get this 400 Bad Request output on the ssh shell.

On the ssh upstream i specified the 8443 port because the service it's listening on a custom port :)
netstat -tunlp |grep 8443
tcp        0      0 0.0.0.0:8443            0.0.0.0:*               LISTEN      26969/sshd: /usr/sb

At the momento i'm not trying to reach an https service via browser, i'm trying just the SSH connection

Thanks again for getting back to me


> [@komodikkio](https://github.com/komodikkio)
> 
> I've created an overview of the route for you. If no suitable domain is specified, your browser will default to "ssh," which it won't be able to do (it won't understand because it's an SSH protocol). "https" upstream is empty and cannot be reached.
> 
> ![Image](https://github.com/user-attachments/assets/10ca53f3-ce8d-4fba-a57e-67abb75a14b3)
> 
> On the upstream "ssh," I see that you're specifying port "8443," which not sounds like the SSH server. Please check this again; the normal SSH port is 22. If there's a web server behind 8443, it won't understand SSH and will respond with a "Bad Request."
> 
> I hope this helps you.

**Comment by stefanwerfling** [comment]
Okay, if you don't want to destroy the NPM 443 config, you'll have to move this server to a different port, 9443. I've made you another diagram. 

![Image](https://github.com/user-attachments/assets/88bbca5a-e2f5-4ff6-8898-738b8b2fba89)

Process: The router redirects the 443 port from the public IP to 9443, which splits the protocol and redirects it internally to 443. 443 is not listed externally (it is not published in Docker). 

```
map $ssl_preread_server_name $name {
     default ssh;
     mysdomain.com https2;
    }

```
$ssl_preread_server_nam You can leave it out, because I'll take care of the names, unless you have a server that issues itself a certificate with Let's Encrypt, in which case you'll need this block. Then you can also point it to your NPM 443 by default.

**Problem:** You now have 9443 as your pre-proxy, so you're losing the client's IP address because the server always shows 127.0.0.1 as the client's IP address. To fix this, you need to enable the proxy protocol. The link tells you what you need to do. But you'll also need to find the 443 config again, where you configure it so that you now receive everything with the proxy protocol header.

https://github.com/NginxProxyManager/nginx-proxy-manager/issues/344#issuecomment-2618540067

good luck! :)

**Comment by komodikkio** [comment]
Thanks @stefanwerfling that would be great.
Sadly i'm on oracle cloud free tier, so i can't do port forwarding but just open/close a port.
I could try to enable the port forwarding on the host itsaelf using iptables
